src/Microsoft.NetCore.Analyzers/Core/Runtime/DoNotUseEnumerableMethodsOnIndexableCollectionsInsteadUseTheCollectionDirectly.Fixer.cs

```diff
@@ -77,14 +79,58 @@ private async Task<Document> UseCollectionDirectly(Document document, TextSpan s
             }
 
             var generator = SyntaxGenerator.GetGenerator(document);
-            var indexNode = generator.LiteralExpression(0);
-            var elementAccessNode = generator.ElementAccessExpression(collectionSyntax.WithoutTrailingTrivia(), indexNode)
-                .WithTrailingTrivia(invocationNode.GetTrailingTrivia());
 
-            var newRoot = root.ReplaceNode(invocationNode, elementAccessNode);
+            var elementAccessNode = GetReplacementNode(methodName, generator, collectionSyntax, semanticModel);
+            if (elementAccessNode == null)
+            {
+                return document;
+            }
+
+            var newRoot = root.ReplaceNode(invocationNode, elementAccessNode.WithTrailingTrivia(invocationNode.GetTrailingTrivia()));
             return document.WithSyntaxRoot(newRoot);
         }
 
+        private static SyntaxNode GetReplacementNode(string methodName, SyntaxGenerator generator, SyntaxNode collectionSyntax, SemanticModel semanticModel)
+        {
+            var collectionSyntaxNoTrailingTrivia = collectionSyntax.WithoutTrailingTrivia();
+
+            if (methodName == "First")
+            {
+                var zeroLiteral = generator.LiteralExpression(0);
+                return generator.ElementAccessExpression(collectionSyntaxNoTrailingTrivia, zeroLiteral);
+            }
+
+            if (methodName == "Last")
+            {
+                // TODO: Handle C# 8 index expression (and vb.net equivalent if any)
+
+                var typeSymbol = semanticModel.GetTypeInfo(collectionSyntax).Type;
+                if (!typeSymbol.HasAnyCollectionCountProperty(WellKnownTypeProvider.GetOrCreate(semanticModel.Compilation)))
+                {
+                    return null;
```

Comment:
Would you be able to refactor the code so the code fix is only registered when we are sure the replacement property exists?

Reply:
Let me know if my attempt looks ok or if it needs some change.